File: services/servico-order/tests/unit/Order.test.ts
import { Order, InvalidOrderStateTransitionError } from '../../src/domain/Order';
import { OrderStatus } from '../../src/domain/OrderStatus';

function makeOrderProps() {
  return {
    consumerId: 'consumer-1',
    restaurantId: 'restaurant-1',
    items: [
      { name: 'Pizza Margherita', quantity: 1, unitPriceCents: 4500 },
      { name: 'Refrigerante', quantity: 2, unitPriceCents: 700 },
    ],
    deliveryAddress: { street: 'Rua das Flores, 123', city: 'São Paulo', zip: '01000-000' },
  };
}

describe('Order (agregado — core domain do FTGO)', () => {
  it('cria um pedido no estado AGUARDANDO_ACEITACAO e publica OrderCreated', () => {
    const { order, event } = Order.create(makeOrderProps());

    expect(order.status).toBe(OrderStatus.AGUARDANDO_ACEITACAO);
    expect(order.totalCents).toBe(4500 + 2 * 700); // 5900
    expect(event.type).toBe('OrderCreated');
    expect(event.aggregateId).toBe(order.id);
    expect(event.payload.totalCents).toBe(5900);
  });

  it('rejeita a criação de um pedido sem itens', () => {
    const props = { ...makeOrderProps(), items: [] };
    expect(() => Order.create(props)).toThrow(/ao menos um item/i);
  });

  it('rejeita item com quantidade inválida', () => {
    const props = makeOrderProps();
    props.items[0].quantity = 0;
    expect(() => Order.create(props)).toThrow(/inválidos/i);
  });

  it('transiciona para PREPARANDO e publica OrderApproved ao receber CreditReserved', () => {
    const { order } = Order.create(makeOrderProps());
    const event = order.onCreditReserved(5900);

    expect(order.status).toBe(OrderStatus.PREPARANDO);
    expect(event.type).toBe('OrderApproved');
    expect(event.payload.reservedAmountCents).toBe(5900);
  });

  it('transiciona para CANCELADO e publica OrderCancelled ao receber CreditRejected', () => {
    const { order } = Order.create(makeOrderProps());
    const event = order.onCreditRejected('Limite insuficiente.');

    expect(order.status).toBe(OrderStatus.CANCELADO);
    expect(order.cancelReason).toBe('Limite insuficiente.');
    expect(event.type).toBe('OrderCancelled');
  });

  it('não permite aprovar um pedido que não está aguardando aceitação', () => {
    const { order } = Order.create(makeOrderProps());
    order.onCreditReserved(5900); // -> PREPARANDO

    expect(() => order.onCreditReserved(5900)).toThrow(InvalidOrderStateTransitionError);
  });

  it('não permite rejeitar crédito de um pedido já cancelado', () => {
    const { order } = Order.create(makeOrderProps());
    order.onCreditRejected('Falha no cartão.');

    expect(() => order.onCreditRejected('Outra falha.')).toThrow(InvalidOrderStateTransitionError);
  });

  it('percorre o fluxo completo até ENTREGUE e publica os eventos esperados, em ordem', () => {
    const { order, event: created } = Order.create(makeOrderProps());
    const approved = order.onCreditReserved(5900);
    order.onTicketPrepared();
    order.onDeliveryScheduled('entregador-42');
    const delivered = order.onDeliveryCompleted();

    expect(order.status).toBe(OrderStatus.ENTREGUE);
    expect(order.courierId).toBe('entregador-42');
    expect(order.deliveredAt).toBeDefined();

    expect([created.type, approved.type, delivered.type]).toEqual([
      'OrderCreated',
      'OrderApproved',
      'OrderDelivered',
    ]);
    expect(delivered.payload.courierId).toBe('entregador-42');
  });

  it('não permite preparar o ticket antes da aprovação de crédito', () => {
    const { order } = Order.create(makeOrderProps());
    expect(() => order.onTicketPrepared()).toThrow(InvalidOrderStateTransitionError);
  });

  it('não permite concluir a entrega sem entregador atribuído', () => {
    const { order } = Order.create(makeOrderProps());
    order.onCreditReserved(5900);
    order.onTicketPrepared();

    expect(() => order.onDeliveryCompleted()).toThrow(/entregador atribuído/i);
  });

  it('permite cancelamento direto antes da entrega, mas não depois', () => {
    const { order } = Order.create(makeOrderProps());
    order.onCreditReserved(5900);
    order.onTicketPrepared();
    order.onDeliveryScheduled('entregador-7');
    order.onDeliveryCompleted();

    expect(() => order.cancel('Consumidor desistiu.')).toThrow(InvalidOrderStateTransitionError);
  });

  it('serializa o pedido via toJSON com os campos esperados', () => {
    const { order } = Order.create(makeOrderProps());
    const json = order.toJSON();

    expect(json).toMatchObject({
      id: order.id,
      status: OrderStatus.AGUARDANDO_ACEITACAO,
      totalCents: 5900,
    });
  });
});

File: services-java/servico-order/src/test/java/br/com/ftgo/order/domain/OrderTest.java
package br.com.ftgo.order.domain;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.util.List;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import br.com.ftgo.order.domain.event.OrderEvents;

class OrderTest {

    private static Order newOrder() {
        return Order.create(
                "consumer-1", "restaurant-1",
                List.of(
                        new OrderItem("Pizza Margherita", 1, 4500),
                        new OrderItem("Refrigerante", 2, 700)
                ),
                new DeliveryAddress("Rua das Flores, 123", "São Paulo", "01000-000"));
    }

    @Test
    @DisplayName("cria pedido em AGUARDANDO_ACEITACAO e calcula o total")
    void createsOrder() {
        Order order = newOrder();
        assertThat(order.getStatus()).isEqualTo(OrderStatus.AGUARDANDO_ACEITACAO);
        assertThat(order.getTotalCents()).isEqualTo(4500 + 2 * 700);

        OrderEvents.OrderCreated event = order.toCreatedEvent();
        assertThat(event.type()).isEqualTo("OrderCreated");
        assertThat(event.aggregateId()).isEqualTo(order.getId());
        assertThat(event.totalCents()).isEqualTo(5900);
    }

    @Test
    @DisplayName("rejeita pedido sem itens")
    void rejectsEmptyItems() {
        assertThatThrownBy(() -> Order.create("c", "r", List.of(),
                new DeliveryAddress("Rua", "Cidade", "00000-000")))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("ao menos um item");
    }

    @Test
    @DisplayName("CreditReserved -> PREPARANDO e publica OrderApproved")
    void creditReserved() {
        Order order = newOrder();
        OrderEvents.OrderApproved event = order.onCreditReserved(5900);
        assertThat(order.getStatus()).isEqualTo(OrderStatus.PREPARANDO);
        assertThat(event.type()).isEqualTo("OrderApproved");
        assertThat(event.reservedAmountCents()).isEqualTo(5900);
    }

    @Test
    @DisplayName("CreditRejected -> CANCELADO e publica OrderCancelled")
    void creditRejected() {
        Order order = newOrder();
        OrderEvents.OrderCancelled event = order.onCreditRejected("Limite insuficiente.");
        assertThat(order.getStatus()).isEqualTo(OrderStatus.CANCELADO);
        assertThat(order.getCancelReason()).isEqualTo("Limite insuficiente.");
        assertThat(event.type()).isEqualTo("OrderCancelled");
    }

    @Test
    @DisplayName("não permite aprovar pedido fora de AGUARDANDO_ACEITACAO")
    void cannotApproveTwice() {
        Order order = newOrder();
        order.onCreditReserved(5900);
        assertThatThrownBy(() -> order.onCreditReserved(5900))
                .isInstanceOf(InvalidOrderStateTransitionException.class);
    }

    @Test
    @DisplayName("fluxo completo até ENTREGUE publica OrderCreated, OrderApproved e OrderDelivered")
    void fullHappyPath() {
        Order order = newOrder();
        var created = order.toCreatedEvent();
        var approved = order.onCreditReserved(5900);
        order.onTicketPrepared();
        order.onDeliveryScheduled("entregador-42");
        var delivered = order.onDeliveryCompleted();

        assertThat(order.getStatus()).isEqualTo(OrderStatus.ENTREGUE);
        assertThat(order.getCourierId()).isEqualTo("entregador-42");
        assertThat(order.getDeliveredAt()).isNotNull();

        assertThat(List.of(created.type(), approved.type(), delivered.type()))
                .containsExactly("OrderCreated", "OrderApproved", "OrderDelivered");
        assertThat(delivered.courierId()).isEqualTo("entregador-42");
    }

    @Test
    @DisplayName("não permite preparar ticket antes da aprovação de crédito")
    void cannotPrepareBeforeApproval() {
        Order order = newOrder();
        assertThatThrownBy(order::onTicketPrepared)
                .isInstanceOf(InvalidOrderStateTransitionException.class);
    }

    @Test
    @DisplayName("não permite concluir entrega sem entregador atribuído")
    void cannotCompleteWithoutCourier() {
        Order order = newOrder();
        order.onCreditReserved(5900);
        order.onTicketPrepared();
        assertThatThrownBy(order::onDeliveryCompleted)
                .isInstanceOf(IllegalStateException.class)
                .hasMessageContaining("entregador");
    }

    @Test
    @DisplayName("não permite cancelar pedido já entregue")
    void cannotCancelAfterDelivered() {
        Order order = newOrder();
        order.onCreditReserved(5900);
        order.onTicketPrepared();
        order.onDeliveryScheduled("entregador-7");
        order.onDeliveryCompleted();
        assertThatThrownBy(() -> order.cancel("Consumidor desistiu."))
                .isInstanceOf(InvalidOrderStateTransitionException.class);
    }
}
